fix recursive & non-recursive kubectl get of generic output format

metral opened the pull request:
This PR fixes the issues with `kubectl get` in https://github.com/kubernetes/kubernetes/issues/26466

Changes made:
- fix printing when using the generic output format in both non-recursive & recurvise settings to ensure that errors are being shown
- add tests to check printing generic output in a **non-recursive** setting with non-existent pods
- clean up the **recursive** `kubectl get` tests

/cc @janetkuo

metral committed:
fix recursive & non-recursive kubectl get of generic output format

smarterclayton commented on pkg/kubectl/cmd/get.go:
```diff
@@ -219,33 +219,55 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
-		obj, err := resource.AsVersionedObject(infos, !singular, version, f.JSONEncoder())
-		if err != nil {
+
+		switch options.Recursive {
```
Recursive should still output a List, not individual items.  So this seems incorrect in that it will result in many lists printed.

janetkuo commented:
ok to test

metral commented:
@janetkuo I've gone ahead and addressed the feedback, PTAL

metral commented on pkg/kubectl/cmd/get.go:
```diff
@@ -315,5 +339,5 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 		}
 	}
-	return utilerrors.Flatten(utilerrors.NewAggregate(allErrs))
+	return utilerrors.NewAggregate(allErrs)
```
@smarterclayton 
> Printing out allErrs as a series of error lines would be consistent with what we do with Create.

By returning an `Aggregate` of the errors to `cmdUtil.CheckErr` after `RunGet` is done, `cmdUtil.CheckErr` ends up printing the series of errors in separate lines  per [MultipleErrors()](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/helpers.go#L139-L145)

Am I missing something?

smarterclayton replied:
You split the code into two chunks (one that uses Visit, and one that doesn't).  The previous code captured errors - nothing prevents them from being printed out without changing the block above to use Visit().

metral replied:
The split you're referring to of using Visit and non-Visit was for the `generic` output case - I've since changed that in my most recent push so that they both make use of `Infos()`.

If I'm understanding you correctly, you're referring to the block of code _after_ the generic case, and how that should be altered to use Visit() to handle the printing correctly. Did I get that right?

smarterclayton replied:
Ok, so looking at the latest my concern is that this is harder to understand than the previous code - all that is different is the IntoSingular(), and you could use a block at the end to check whether this was singular before returning.  I think collapsing them back down into one would be clearer.

Looking at this more,  I just noticed that you were keying on Recursive - but Recursive should be a detail inside of resource builder, and specifying recursive should automatically set singular to false (inside of the resource.Builder code).  The intent of singular was to resolve all the inputs down into "did the user give us a single object or not" and passing recursive is a definite "not".

metral commented:
@smarterclayton 

> I think collapsing them back down into one would be clearer.

> Looking at this more, I just noticed that you were keying on Recursive - but Recursive should be a detail inside of resource builder, and specifying recursive should automatically set singular to false (inside of the resource.Builder code).

I've addressed your feedback and squashed down, PTAL

metral commented:
@smarterclayton 

> Mutating b.singular is not appropriate - you should initialize isSigular and then mutate that. Any call that returns a Result should be side effect free.

all set, ptal

metral commented:
should this PR have a v1.3 milestone since https://github.com/kubernetes/kubernetes/issues/26466 is both a bug and a P2?

smarterclayton commented:
Yes, since it was eating errors.

metral commented:
Jenkins GCE e2e is hitting a flake https://github.com/kubernetes/kubernetes/issues/26760

metral commented:
squashed down

metral commented:
@smarterclayton @janetkuo can I get the LGTM again? I had to squash down

Thanks.

metral commented:
@janetkuo can this PR get an LGTM to move things along? I had to squash down so the label was removed - thanks!

metral commented:
@janetkuo The submit queue tests are hitting flake https://github.com/kubernetes/kubernetes/issues/26210

Can we issue a retest to get it back into the queue? Thanks

janetkuo commented:
@k8s-bot e2e test this issue: #26210

ncdc commented:
FYI, prior to #23673 and this PR, `kubectl get pod/notfound -o yaml` would print out an error. Now, what I'm seeing on master is:

```
$ kubectl get pod/notfound -o yaml
apiVersion: v1
items: []
kind: List
metadata: {}
Error from server: pods "foo" not found
```

ncdc commented:
That's because this doesn't return the error immediately: https://github.com/kubernetes/kubernetes/blob/760b04e294d82cd2d2d0ff5a32e8999b34d3f229/pkg/kubectl/cmd/get.go#L235
